zipline/data/fx/hdf5.py

```diff
@@ -0,0 +1,271 @@
+row i in a data node is the ith element of /index/dts.
+"""
+from interface import implements
+import h5py
+from logbook import Logger
+import pandas as pd
+import six
+
+from zipline.utils.memoize import lazyval
+
+from .base import FXRateReader
+
+INDEX = 'index'
+DATA = 'data'
+CURRENCIES = 'currencies'
+DTS = 'dts'
+RATES = 'rates'
+DEFAULT = 'default'
+
+log = Logger(__name__)
+
+
+class HDF5FXRateReader(implements(FXRateReader)):
+    """An FXRateReader backed by HDF5.
+
+    Parameters
+    ----------
+    group : h5py.Group
+        Top-level group written by an :class:`HDF5FXRateWriter`.
+    default_rate : str
+        Rate to use when ``get_rates`` is called with a rate of 'default'.
+    """
+
+    def __init__(self, group, default_rate):
+        self._group = group
+        self._default_rate = default_rate
+
+    @classmethod
+    def from_path(cls, path):
+        """Construct from a file path.
+
+        Parameters
+        ----------
+        path : str
+            Path to an HDF5 fx rates file.
+        """
+        return cls(h5py.File(path))
+
+    @lazyval
+    def dts(self):
+        """Row labels for rate groups.
+        """
+        return pd.DatetimeIndex(
+            self._group[INDEX][DTS][:].astype('M8[ns]'),
+            tz='UTC',
+        )
+
+    @lazyval
+    def currencies(self):
+        """Column labels for rate groups.
+        """
+        # Currencies are stored as fixed-length bytes in the file, but we want
+        # `str` objects in memory.
+        byte_strings = self._group[INDEX][CURRENCIES][:]
+        if six.PY3:
+            values = [c.decode('ascii') for c in byte_strings]
+        else:
+            values = byte_strings.astype(object)
+
+        return pd.Index(values)
+
+    def get_rates(self, rate, quote, bases, dts):
+        """Get rates to convert ``bases`` into ``quote``.
+        """
+        if rate == 'default':
+            rate = self._default_rate
+
+        self._check_dts(self.dts, dts)
+
+        row_ixs = self.dts.searchsorted(dts, side='right') - 1
+        col_ixs = self.currencies.get_indexer(bases)
+
+        return self._read_rate_block(rate, quote, row_ixs, col_ixs)
+
+    def _read_rate_block(self, rate, quote, row_ixs, col_ixs):
+        try:
+            dataset = self._group[DATA][rate][quote][RATES]
+        except KeyError:
+            raise ValueError(
+                "FX rates not available for rate={}, quote_currency={}."
+                .format(rate, quote)
+            )
+
+        # There aren't many columns in the output array, so it's easier to pull
+        # all columns and reindex in memory. For rows, however, a quick and
+        # easy optimization is to pull just the slice from min(row_ixs) to
+        # max(row_ixs).
+        min_row = row_ixs[0]
+        max_row = row_ixs[-1]
+        rows = dataset[min_row:max_row + 1]  # +1 to be inclusive of end
+
+        return rows[row_ixs - min_row][:, col_ixs]
```

Comment:
I'm having trouble keeping track of what the rows and cols are. Worth renaming `row` to `date`, and `col` to `currency`? (Are those interpretations actually correct?)